### PR TITLE
[FrameworkBundle] Add trans method shortcut to controllers

### PR DESCRIPTION
**Pull request by Babacooll**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |
- [ ] Add related documentation

Add simple trans method to controllers.

### CODE REVIEW

**Comment by Tobion** [comment]
:-1: what about transChoice?

Also the commit contains files that should not be commited.

**Comment by Babacooll** [comment]
Indeed, I've fixed those commited files.

Do you think transChoice should be added too as a shorcut?

**Comment by Babacooll** [comment]
I added the transChoice shortcut

**Comment by aitboudad** [comment]
@Tobion I rarely use trans in controller but it's useful in some case like when using ajax call.

**Comment by fabpot** [comment]
I'm :-1:. Methods in the base controller should be used by the vast majority of our users.

Maybe it's time to add some traits like we've done in Silex.

**Comment by aitboudad** [comment]
adding traits, sound good to me :)

**Comment by Babacooll** [comment]
I do agree about the fact it's not something widely used. The FrameworkBundle Controller solution seems quite inefficient in this case.

A solution based on trait like you suggested seems a better way.

Do you want me to adapt it into a Trait like done in Silex ?

**Comment by aitboudad** [comment]
@Babacooll I think so

**Comment by Babacooll** [comment]
Here are the modifications using the Trait and the related unit tests.

But after reflexion I think the integration of the addFlash shortcut should imply the integration of those two shortcuts in the Controller as flash messages will often be translated.

Plus I don't think the TranslationTrait can be reused in other classes than controllers as injecting the Container is a bad practice.

File: src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace Symfony\Bundle\FrameworkBundle\Application;
+
+/**
+ * Translation trait.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ * @author Michaël Garrez <michael.garrez@gmail.com>
+ */
+trait TranslationTrait
```
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php]: `TranslatorTrait`

File: src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+     * @param string|null $domain     The domain for the message or null to use the default
+     * @param string|null $locale     The locale or null to use the default
+     *
+     * @throws \InvalidArgumentException If the locale contains invalid characters
+     *
+     * @return string The translated string
+     */
+    public function trans($id, array $parameters = array(), $domain = null, $locale = null)
+    {
+        if (!$this->container->has('translator')) {
```
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php]: it is not necessary

File: src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Application;
```
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php]: can we move it into `Symfony\Bundle\FrameworkBundle\Translation` namespace

File: src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+     * @param string|null $domain     The domain for the message or null to use the default
+     * @param string|null $locale     The locale or null to use the default
+     *
+     * @throws \InvalidArgumentException If the locale contains invalid characters
+     *
+     * @return string The translated string
+     */
+    public function transChoice($id, $number, $parameters = array(), $domain = null, $locale = null)
+    {
+        if (!$this->container->has('translator')) {
```
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php]: same here

File: src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            throw new \LogicException('You can not use the trans method if translator is disabled.');
+        }
+
+        return $this->container->get('translator')->trans($id, $parameters, $domain, $locale);
+    }
+
+    /**
+     * Translates the given choice message by choosing a translation according to a number.
```
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php]: we can use 
```php
 /**
  * @see TranslatorInterface::trans()
  */
```

File: src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+/**
+ * Translation trait.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ * @author Michaël Garrez <michael.garrez@gmail.com>
+ */
+trait TranslationTrait
+{
+    /**
+     * Translates the given message.
```
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php]:
```php
 /**
  * @see TranslatorInterface::trans()
  */

File: src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php (hunk 1 above)
Comment (Babacooll) [inline src/Symfony/Bundle/FrameworkBundle/Application/TranslationTrait.php]: I did use the same name as the Silex TranslationTrait, don't you think we should use the same pattern?

**Comment by Babacooll** [comment]
I fixed some of your comments except for the test checking if the container contains the translator.

Is it impossible to disable the Translator or remove its access from the Container ?

For the trait name, I used the same as Silex TranslationTrait.

**Comment by dosten** [comment]
If this gets accepted i think we should add a `SecurityTrait`, `RouterTrait` too.

**Comment by aitboudad** [comment]
@Babacooll IMO checking the container is useless here 
For the trait name I'd prefer `TranslatorTrait` to make it consistent with service name `translator`

**Comment by aitboudad** [comment]
@Babacooll can you add a note in [CHANGELOG](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md) file

**Comment by Babacooll** [comment]
@aitboudad I removed the check over the container as asked and updated the CHANGELOG

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * Added possibility to extract translation messages from a file or files besides extracting from a directory
  * Added `TranslationsCacheWarmer` to create catalogues at warmup
+ * Added `TranslatorTrait` which provides shortcuts for Translator `trans()` and `transChoice()` methods
```
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: it should be added in `3.0.0` section.

**Comment by aitboudad** [comment]
:+1:

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md (hunk 7 above)
Comment (Babacooll) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: My bad, fixed !

**Comment by Tobion** [comment]
I don't agree with the current implementation. The TranslatorTrait uses the container, but the container is not known by the trait. So either we

- add a ContainerTrait, which introduces the container property (and container setter) and this trait is then used by the TranslatorTrait
- or the Translator does not access the container and instead introduces a translator property which is then accessed instead of the container

Also if we add traits, then all of the base controller functionality should be changed/added as trait.

**Comment by aitboudad** [comment]
@Tobion the current implementation is expected to be called only in controller.
 I think the second option is better then when using controller as service for example

**Comment by Babacooll** [comment]
If we want this Trait to be used it must be as simple as possible, the second option seems a better implementation IMO. I'm waiting for more feedbacks and I will do the update.

**Comment by Babacooll** [comment]
@Tobion  @aitboudad  Any updates about the solution you think is preferable?

**Comment by aitboudad** [comment]
IMO I would prefer to use the same way as Silex based on the Container if we use @Tobion proposition we need Injecting tons of services into controllers !!

**Comment by Babacooll** [comment]
@aitboudad @Tobion Do you think a defensive check on the implementation of ContainerAwareInterface could solve this issue? We can then assure the class implementing the Trait has access to the container

**Pull request closed without merging**